### ISSUE TITLE
dataframe.map_partitions wrong usage or? mapreduce

### ISSUE THREAD

**Issue by mfouesneau** [dask/dask#1541]
I cannot find the answer anywhere sorry.

I try to map a function onto each partition of a dataframe. 
I have a dataframe with 30 partitions and I try for instance 

```python
d.map_partitions(len, meta=int)
```
but it only return the last partition size, while `len(d)` returns the correct result
Of course I have more complex functions to apply.

I tried to print or count the calls to the function to map and I never get the number of partitions.

What's the correct way to apply on each partition a function? 
basically mapreduce some dataframe.

**Comment by jcrist** [comment]
- `meta` isn't the metadata for your function, it's the metadata for *the entire output*. In this case, the desired output is a series, so you should pass in an empty series with integer dtype, or `(series_name, int)` to have that created automatically for you.
- Unless you're computing immediately afterwards, it's highly advisable to have the mapped function return a series/dataframe/index. `dask.dataframe` expects each partition of the data to be a pandas type, but in this case you're returning an integer.
- The `reduction` method may be actually what you're looking for. See docs here: http://dask.pydata.org/en/latest/dataframe-api.html#dask.dataframe.DataFrame.reduction. This will do the coercion to pandas objects for you, as well as manage the map and reduce steps.

**Comment by mrocklin** [comment]
Agreed with @jcrist.

In the future it might be nice either to be robust to scalar outputs/meta
or to err more loudly.

**Comment by mfouesneau** [comment]
That helps, slightly... I indeed got the simple `len` working, but I do not have the reduction method. Which version is implementing this?? I have 0.11.0

**Comment by jcrist** [comment]
Looks like it was added just after the last release (https://github.com/dask/dask/pull/1483). If possible, I recommend using the master branch:

`pip install -U git+https://github.com/dask/dask.git`

**Comment by mfouesneau** [comment]
I managed to get the master branch. Indeed this method appears
How do you use it properly?

this works
`d.reduction(len, aggregate=sum).compute()`

but I it is unclear how I can return arrays or complex data from the mapped function. I could not return tuple properly just based on the docs.

**Comment by jcrist** [comment]
Without knowing more about what you're trying to do, it's hard to say. The reduction method does the following:
- Apply `chunk` to each partition
- Merge the chunks (in a way that makes sense for the types involved)
- Apply `aggregate` to the merged result

If you're not passing around dataframes, series, or scalars, then you might find working with `dask.delayed` (http://dask.pydata.org/en/latest/delayed.html) to be a better fit. `dask.dataframe` is designed for working on collections where each partition is a pandas object, not arbitrary things.

```python
parts = df.to_delayed()  # List of Delayed objects, one for each partition

@delayed(pure=True)
def my_chunk(partition):
    ...

@delayed(pure=True)
def my_agg(list_of_chunks):
    ...

chunks = map(my_chunk, parts)
result = my_agg(chunks)
```

**Comment by mfouesneau** [comment]
I use a large collection of csv on which I can do map reduce like computations.

Your solution seems to produce the right tasks, (`result.visualize()`) but then I get an error 
`TypeError: map() must have at least two arguments.`

Also what's the use of `pure`?

**Comment by mfouesneau** [comment]
I found the `map` issue. In python 3, map returns generators not lists. Apparently this does not suit the delayed function.

**Comment by jcrist** [comment]
> Also what's the use of pure? 

From the docs (http://dask.pydata.org/en/latest/delayed-api.html#dask.delayed.delayed):

```
pure : bool, optional
    Indicates whether calling the resulting ``Delayed`` object is a pure
    operation. If True, arguments to the call are hashed to produce
    deterministic keys. Default is False.
```

If you don't set `pure=True`, then calling your function multiple times on the same inputs will create different nodes each time, rather than sharing the result of one call. There's an explanation of the rationale for this in the docstring of `delayed` (along with an example) - if it doesn't answer your question please let us know how we can improve it.

**Comment by mfouesneau** [comment]
That's great thanks. I did not see that it would create different nodes but it is probably that I need to digest more.  I do not fully see why you would like different nodes but that's ok.

Thanks Dask will help me in many tasks.

**Comment by jcrist** [comment]
Taking the example from the docs:

```python
>>> from random import random
>>> out1 = delayed(random, pure=False)()
>>> out2 = delayed(random, pure=False)()
>>> out1.key == out2.key
False
```

If `pure=True` was used, then calling `compute(out1, out2)` would result in the same number for both calls to random, as dask would only call `random` once (instead of twice). This is because functions that are marked as pure (the output only depends on the input) have their input arguments hashed, resulting in the same key names on every call:

```python
>>> @delayed(pure=True)
... def add(a, b):
...     return a + b
>>> out1 = add(1, 2)
>>> out2 = add(1, 2)
>>> out1.key == out2.key
True
```

Usually this behavior is nice, as it helps to avoid repeating work. However, some functions may not be "pure" (such as `random`, where the output *doesn't* depend on the input alone). To be safe, `pure=False` by default, so this kind of subexpression-elimination doesn't happen unexpectedly.

**Comment by jcrist** [comment]
Did you have any other questions, or can this issue be closed?

**Comment by mfouesneau** [comment]
I did not think about have some random components inside functions. Makes sense.
We can close the issue. Thanks for the help!

**Comment by CMCDragonkai** [comment]
Given that `map_partitions` takes each partition and returns another partition (where this partition is bigger or smaller), it appears that `map_partitions` is capable of squeezing the total length of the dataframe or increasing the total length of the data frame.
